base/loading.jl

```diff
@@ -1159,59 +1158,84 @@ end
-                    else
-                        pkg, env = pkgenv
-                        path = Base.locate_package(pkg, env)
-                        if path === nothing
-                            ext_not_allowed_load = trigger
-                            break
-                        end
-                    end
-                end
-                if ext_not_allowed_load !== nothing
-                    @debug "Extension $(extid.id.name) of $(extid.parentid.name) not loaded due to \
-                            $(ext_not_allowed_load.name) loaded in environment lower in load path"
-                else
-                    require(extid.id)
-                    @debug "Extension $(extid.id.name) of $(extid.parentid.name) loaded"
-                end
-                extid.succeeded = true
-            end
-        end
+function run_extension_callbacks(extid::ExtensionId)
+    assert_havelock(require_lock)
+    succeeded = try
+        _require_prelocked(extid.id)
+        @debug "Extension $(extid.id.name) of $(extid.parentid.name) loaded"
+        true
     catch
         # Try to continue loading if loading an extension errors
         errs = current_exceptions()
         @error "Error during loading of extension" exception=errs
+        false
+    end
+    return succeeded
+end
+
+function run_extension_callbacks(pkgid::PkgId)
+    assert_havelock(require_lock)
+    # take ownership of extids that depend on this pkgid
+    extids = pop!(EXT_DORMITORY, pkgid, nothing)
+    extids === nothing && return
+    for extid in extids
+        if extid.ntriggers > 0
+            # It is possible that pkgid was loaded in an environment
+            # below the one of the parent. This will cause a load failure when the
+            # pkg ext tries to load the triggers. Therefore, check this first
+            # before loading the pkg ext.
+            pkgenv = Base.identify_package_env(extid.id, pkgid.name)
+            ext_not_allowed_load = false
+            if pkgenv === nothing
+                ext_not_allowed_load = true
+            else
+                pkg, env = pkgenv
+                path = Base.locate_package(pkg, env)
+                if path === nothing
+                    ext_not_allowed_load = true
+                end
+            end
+            if ext_not_allowed_load
+                @debug "Extension $(extid.id.name) of $(extid.parentid.name) will not be loaded \
+                        since $(pkgid.name) loaded in environment lower in load path"
+                # indicate extid is expected to fail
+                extid.ntriggers *= -1
+            else
+                # indicate pkgid is loaded
+                extid.ntriggers -= 1
+            end
+        end
+        if extid.ntriggers < 0
+            # indicate pkgid is loaded
+            extid.ntriggers += 1
+            succeeded = false
+        else
+            succeeded = true
+        end
+        if extid.ntriggers == 0
```

Comment:
This should also check that the extension is not in `EXT_DORMITORY_FAILED`?

Reply:
Maybe something like

```diff
diff --git a/base/loading.jl b/base/loading.jl
index 25c89c31ca..e57e2e530a 100644
--- a/base/loading.jl
+++ b/base/loading.jl
@@ -1105,7 +1105,7 @@ mutable struct ExtensionId
 end
 
 const EXT_DORMITORY = Dict{PkgId,Vector{ExtensionId}}()
-const EXT_DORMITORY_FAILED = ExtensionId[]
+const EXT_DORMITORY_FAILED = Dict{PkgId,ExtensionId}()
 
 function insert_extension_triggers(pkg::PkgId)
     pkg.uuid === nothing && return
@@ -1229,11 +1229,11 @@ function run_extension_callbacks(pkgid::PkgId)
         else
             succeeded = true
         end
-        if extid.ntriggers == 0
+        if extid.ntriggers == 0 && !haskey(EXT_DORMITORY_FAILED, extid.id)
             # actually load extid, now that all dependencies are met,
             # and record the result
             succeeded = succeeded && run_extension_callbacks(extid)
-            succeeded || push!(EXT_DORMITORY_FAILED, extid)
+            succeeded || (EXT_DORMITORY_FAILED[extid.id] = extid)
         end
     end
     nothing
@@ -1253,10 +1253,10 @@ function retry_load_extensions()
     # so this can still mutate the list to drop successful ones
     failed = copy(EXT_DORMITORY_FAILED)
     empty!(EXT_DORMITORY_FAILED)
-    filter!(failed) do extid
+    filter!(failed) do (_, extid)
         return !run_extension_callbacks(extid)
     end
-    prepend!(EXT_DORMITORY_FAILED, failed)
+    merge!(EXT_DORMITORY_FAILED, failed)
     end
     nothing
 end
```

Reply:
That could be an interesting change, though I don't think it is strictly related to this PR. It only changes the case where something failed failed to load, while previously I was not intending to prevent a package from registering the same extension multiple times, possibly with different trigger sets.